### PR TITLE
Fix potential NPE in UIViewOperationQueue.

### PR DESCRIPTION
**Pull request by manicakes**
We were noticing the following crash in our application, that was occurring fairly often, but still hard to reproduce:

```
12-12 10:37:35.342: E/AndroidRuntime(9064): FATAL EXCEPTION: main
12-12 10:37:35.342: E/AndroidRuntime(9064): Process: com.bloomberg.android.plus, PID: 9064
12-12 10:37:35.342: E/AndroidRuntime(9064): java.lang.NullPointerException: Attempt to invoke interface method 'void com.facebook.react.uimanager.UIViewOperationQueue$UIOperation.execute()' on a null object reference
12-12 10:37:35.342: E/AndroidRuntime(9064): 	at com.facebook.react.uimanager.UIViewOperationQueue$2.run(UIViewOperationQueue.java:782)
12-12 10:37:35.342: E/AndroidRuntime(9064): 	at com.facebook.react.uimanager.UIViewOperationQueue.flushPendingBatches(UIViewOperationQueue.java:829)
12-12 10:37:35.342: E/AndroidRuntime(9064): 	at com.facebook.react.uimanager.UIViewOperationQueue.access$1500(UIViewOperationQueue.java:44)
12-12 10:37:35.342: E/AndroidRuntime(9064): 	at com.facebook.react.uimanager.UIViewOperationQueue$DispatchUIFrameCallback.doFrameGuarded(UIViewOperationQueue.java:868)
12-12 10:37:35.342: E/AndroidRuntime(9064): 	at com.facebook.react.uimanager.GuardedChoreographerFrameCallback.doFrame(GuardedChoreographerFrameCallback.java:32)
12-12 10:37:35.342: E/AndroidRuntime(9064): 	at com.facebook.react.uimanager.ReactChoreographer$ReactChoreographerDispatcher.doFrame(ReactChoreographer.java:131)
12-12 10:37:35.342: E/AndroidRuntime(9064): 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:872)
12-12 10:37:35.342: E/AndroidRuntime(9064): 	at android.view.Choreographer.doCallbacks(Choreographer.java:686)
12-12 10:37:35.342: E/AndroidRuntime(9064): 	at android.view.Choreographer.doFrame(Choreographer.java:618)
12-12 10:37:35.342: E/AndroidRuntime(9064): 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:860)
12-12 10:37:35.342: E/AndroidRuntime(9064): 	at android.os.Handler.handleCallback(Handler.java:751)
12-12 10:37:35.342: E/AndroidRuntime(9064): 	at android.os.Handler.dispatchMessage(Handler.java:95)
12-12 10:37:35.342: E/AndroidRuntime(9064): 	at android.os.Looper.loop(Looper.java:154)
12-12 10:37:35.342: E/AndroidRuntime(9064): 	at android.app.ActivityThread.main(ActivityThread.java:6088)
12-12 10:37:35.342: E/AndroidRuntime(9064): 	at java.lang.reflect.Method.invoke(Native Method)
12-12 10:37:35.342: E/AndroidRuntime(9064): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
12-12 10:37:35.342: E/AndroidRuntime(9064): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
```

Looks like we may have `null` objects in the `UIViewOperationQueue`. I went ahead and put in null checks in the relevant places.

### CODE REVIEW

**Comment by gre** [comment]
looks good to me

**Comment by astreet** [comment]
This would imply that somewhere we're adding null to the queue. A better fix would be to never allow null to be added to the queue -- there's no reason I can think of that we'd ever want that.

**Comment by manicakes** [comment]
@astreet makes sense, though a bit time constrained at the moment, probably won't be able to find the root cause anytime soon. Shall I close this PR and open an issue?

**Comment by astreet** [comment]
There's nothing that I can see that could add null to the operations queue in open source -- do you have any non-open source usages of UIViewOperationQueue#enqueueUIOperation? All other add calls I can see add a new object that can't be null. I'm worried we might be hiding a bug somewhere (in our code or yours) or could hide a bug in the future.

**Comment by manicakes** [comment]
@astreet I think this is worth investigating. Our code is unfortunately not open source, but as soon as I can reproduce this consistently, I will either fix the bug or put together a sample app that causes the error.

**Comment by manicakes** [comment]
So weird - I'm not seeing *any* calls to `enqueueUIOperation` other than the method definition. Grepping our entire source tree, including imported modules:

```
$ find . -type f -name "*.java" -exec grep -il enqueueUIOperation {} \;
./node_modules/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIViewOperationQueue.java
```

**Comment by astreet** [comment]
Alright, thanks for following up. Not saying it can't somehow be a bug in OSS, but we haven't had this crop up for any of our apps at FB AFAICT

**Comment by manicakes** [comment]
NP - will try to spend some time on this today and hopefully will find something. Cheers.

**Comment by manicakes** [comment]
@astreet unfortunately didn't get to the bottom of this. However, I updated this PR to include a null check for the `enqueueUIOperation` method; thinking that should be there anyway. Whether it fixes the NPE is another matter. :)

**Comment by manicakes** [comment]
@astreet any thoughts on this PR? This is currently our most common crash on Android, so your help is much appreciated.

**Comment by astreet** [comment]
Instead of a null-check, I'm fine adding `SoftAssertions.assertNotNull` to enqueueUIOperation: this should catch the problem earlier and give you a better stack trace.

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/UIViewOperationQueue.java [hunk 1]
```diff
@@ -778,8 +780,8 @@ public void run() {
                  }
 
                  if (operations != null) {
-                   for (int i = 0; i < operations.size(); i++) {
```
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/UIViewOperationQueue.java]: Remove this change since it's not relevant.

**Comment by manicakes** [comment]
@astreet updated, thanks!

**Comment by astreet** [comment]
@facebook-github-bot shipit

**Comment by astreet** [comment]
Thanks!

**Comment by dmueller39** [comment]
@astreet After reexamining this problem this issue isn't due to adding null to mOperations, its due to mutating mOperations from the main UI thread. The SoftAssertion here probably isn't doing anything. I'd like to add SOMETHING to help people in the future, but I'm not certain what.

**Comment by dmueller39** [comment]
Actually the issues isn't the main thread, its any thread other than the native_modules thread. asserting on `UiThreadUtil.isOnUiThread()` would catch 95% of cases, but not 100% of cases (if say, someone spins up their own thread) But given that this problem seems to be pretty uncommon, maybe its enough.

**Comment by AndrewMorsillo** [comment]
This is happening to me @dmueller39 and It's driving me mad. I opened https://github.com/facebook/react-native/issues/22867. I see no way that mOperations can get a null reference put into it. I searched my whole node_modules looking for things that call enqueueUIOperation and there are none. This crash is happening for lots of my users =(

**Comment by AndrewMorsillo** [comment]
@manicakes I had to fork and re-add your change of if(op !== null) since things were still crashing frequently without it. As @dmueller39 said the SoftAssertion doesn't appear to do anything. I still haven't been able to figure out the root cause of the issue, but my app seems to run fine with the extra null check.

**Comment by devwasabi2** [comment]
> Actually the issues isn't the main thread, its any thread other than the native_modules thread. asserting on `UiThreadUtil.isOnUiThread()` would catch 95% of cases, but not 100% of cases (if say, someone spins up their own thread) But given that this problem seems to be pretty uncommon, maybe its enough.

@dmueller39 @evelant @astreet I actually came across a similar bug (where there's a null value in the mOperations queue). It turns out that we are also mutating the queue using enqueueUIBlock. I've been able to prove that we are calling the enqueueUIBlock from the Javascript Queue thread, not from the native modules queue thread. But what I wanna know is how you were able to prove/conclude that mutating the queue from any thread other than the native_modules thread would cause race conditions?

**Pull request closed without merging**